File: EternalTriangle/EternalTriangle/Sequencer.m
//
//  Sequencer.m
//  EternalTriangle
//
//  Created by hokada on 9/12/15.
//  Copyright (c) 2015 Haruki Okada. All rights reserved.
//

#import "Sequencer.h"
#import <EternalTriangle/EternalTriangle-Swift.h>
#import <AudioToolbox/AudioToolbox.h>

#pragma C function prototype decls
static void checkError(OSStatus);
static OSStatus audioUnitCallback(void *,
                                  AudioUnitRenderActionFlags*,
                                  const AudioTimeStamp*,
                                  UInt32,
                                  UInt32,
                                  AudioBufferList*);

static void prepareAUGraph(AUGraph*, AUNode*, AudioUnit*);
static void loadSF2Preset(UInt8, AudioUnit);
static MusicTrack createTrack(MusicSequence, AUNode);
static MusicSequence createSequence(AUGraph);
static MusicPlayer createPlayer(MusicSequence);

#pragma Class body

@interface Sequencer ()

@property (nonatomic) AUGraph processingGraph;
@property (nonatomic) AudioUnit mixerUnit;
@property (nonatomic) AUNode mixerNode;
@property (nonatomic) MusicPlayer musicPlayer;
@property (nonatomic) MusicSequence musicSequence;

@property (nonatomic) UInt32 numTracks;

@end

@implementation Sequencer

- (instancetype)init {
  self = [super init];

  if (self) {
    self.numTracks = 2;

    prepareAUGraph(&_processingGraph, &_mixerNode, &_mixerUnit);
    self.musicSequence = createSequence(_processingGraph);
    self.musicPlayer = createPlayer(_musicSequence);

    [self prepareSamplerNodes];
    [self startAUGraph];
  }

  return self;
}

- (void)dealloc {

}

- (void)play {
  Boolean playing = 0;
  checkError(MusicPlayerIsPlaying(_musicPlayer, &playing));
  if (playing) {
    checkError(MusicPlayerStop(_musicPlayer));
  }

  checkError(MusicPlayerSetTime(_musicPlayer, 0));
  checkError(MusicPlayerStart(_musicPlayer));
}

- (void)stop {
  checkError(MusicPlayerStop(_musicPlayer));
}

- (void)prepareSamplerNodes {
  checkError(AudioUnitSetProperty(_mixerUnit,
                                  kAudioUnitProperty_ElementCount,
                                  kAudioUnitScope_Input,
                                  0,
                                  &_numTracks, sizeof(_numTracks)));

  for (UInt32 i = 0; i < _numTracks; i++) {
    // create sampler
    AUNode samplerNode;
    AudioComponentDescription desc;
    desc.componentType = kAudioUnitType_MusicDevice;
    desc.componentSubType = kAudioUnitSubType_Sampler;
    desc.componentManufacturer = kAudioUnitManufacturer_Apple;
    desc.componentFlags = 0;
    desc.componentFlagsMask = 0;
    checkError(AUGraphAddNode(_processingGraph, &desc, &samplerNode));

    AudioUnit samplerUnit;
    checkError(AUGraphNodeInfo(_processingGraph, samplerNode, NULL, &samplerUnit));
    checkError(AUGraphConnectNodeInput(_processingGraph, samplerNode, 0, _mixerNode, i));

    loadSF2Preset(0, samplerUnit);
    createTrack(_musicSequence, samplerNode);
  }
}

- (void)startAUGraph {
  Boolean isInitialized = 0;
  checkError(AUGraphIsInitialized(_processingGraph, &isInitialized));
  if (isInitialized) {
    checkError(AUGraphInitialize(_processingGraph));
  }

  Boolean isRunning = 0;
  checkError(AUGraphIsRunning(_processingGraph, &isRunning));
  if (isRunning) {
    checkError(AUGraphStart(_processingGraph));
  }
}

@end

#pragma C function decls
static void checkError(OSStatus err) {
  NSCAssert(err == noErr, @"error code : %@", @(err));
}

static OSStatus audioUnitCallback(void *inRefCon,
                                  AudioUnitRenderActionFlags*	ioActionFlags,
                                  const AudioTimeStamp *inTimeStamp,
                                  UInt32 inBusNumber,
                                  UInt32 inNumberFrames,
                                  AudioBufferList *ioData) {
  NSLog(@"called. inTimeStamp : %@", @(inTimeStamp->mSampleTime));
  return noErr;
}

static void prepareAUGraph(AUGraph *outGraph, AUNode *mixerNode, AudioUnit *mixerUnit) {
  // create AUGraph
  checkError(NewAUGraph(outGraph));
  checkError(AUGraphOpen(*outGraph));

  // create ioNode
  AUNode ioNode;
  AudioUnit ioUnit;
  AudioComponentDescription ioUnitDesc;
  ioUnitDesc.componentType = kAudioUnitType_Output;
  ioUnitDesc.componentSubType = kAudioUnitSubType_RemoteIO;
  ioUnitDesc.componentManufacturer = kAudioUnitManufacturer_Apple;
  ioUnitDesc.componentFlags = 0;
  ioUnitDesc.componentFlagsMask = 0;
  checkError(AUGraphAddNode(*outGraph, &ioUnitDesc, &ioNode));
  checkError(AUGraphNodeInfo(*outGraph, ioNode, NULL, &ioUnit));

  // create mixerNode
  AudioComponentDescription mixerDesc;
  mixerDesc.componentType = kAudioUnitType_Mixer;
  mixerDesc.componentSubType = kAudioUnitSubType_MultiChannelMixer;
  mixerDesc.componentManufacturer = kAudioUnitManufacturer_Apple;
  mixerDesc.componentFlags = 0;
  mixerDesc.componentFlagsMask = 0;
  checkError(AUGraphAddNode(*outGraph, &mixerDesc, mixerNode));
  checkError(AUGraphNodeInfo(*outGraph, *mixerNode, NULL, mixerUnit));

  // set callback
  checkError(AudioUnitAddRenderNotify(*mixerUnit, audioUnitCallback, NULL));

  AudioUnitElement ioUnitOutputElement = 0;
  AudioUnitElement mixerOutputElement = 0;
  checkError(AUGraphConnectNodeInput(*outGraph, *mixerNode, mixerOutputElement, ioNode, ioUnitOutputElement));
}

static void loadSF2Preset(UInt8 preset, AudioUnit unit) {
  NSURL* bankURL = [[NSBundle bundleForClass:[Sequencer class]] URLForResource:@"vibraphone" withExtension:@"sf2"];
  AUSamplerInstrumentData instData;
  instData.fileURL = (__bridge CFURLRef)bankURL;
  instData.instrumentType = kInstrumentType_DLSPreset;
  instData.bankMSB = kAUSampler_DefaultMelodicBankMSB;
  instData.bankLSB = kAUSampler_DefaultBankLSB;
  instData.presetID = preset;

  checkError(AudioUnitSetProperty(unit,
                                  kAUSamplerProperty_LoadInstrument,
                                  kAudioUnitScope_Global,
                                  0,
                                  &instData,
                                  sizeof(AUSamplerInstrumentData)));
}

static MusicTrack createTrack(MusicSequence sequence, AUNode inNode) {
  MusicTrack track;
  checkError(MusicSequenceNewTrack(sequence, &track));

  MusicTimeStamp beat = 1.0;
  for (UInt8 i = 60; i <= 72; i++) {
    MIDINoteMessage message;
    message.channel = 0;
    message.note = i;
    message.velocity = 64;
    message.releaseVelocity = 0;
    message.duration = 0.5;

    checkError(MusicTrackNewMIDINoteEvent(track, beat, &message));
    beat += 0.5;
  }

  checkError(MusicTrackSetDestNode(track, inNode));

  return track;
}

static MusicSequence createSequence(AUGraph inGraph) {
  MusicSequence musicSequence;
  checkError(NewMusicSequence(&musicSequence));
  checkError(MusicSequenceSetAUGraph(musicSequence, inGraph));

  return musicSequence;
}

static MusicPlayer createPlayer(MusicSequence inSequence) {
  MusicPlayer player;
  checkError(NewMusicPlayer(&player));
  checkError(MusicPlayerSetSequence(player, inSequence));
  checkError(MusicPlayerPreroll(player));

  return player;
}

File: EternalTriangle/EternalTriangle/Sequencer.swift
//
//  Sequencer2.swift
//  EternalTriangle
//
//  Created by hokada on 10/30/15.
//  Copyright © 2015 Haruki Okada. All rights reserved.
//

import Foundation
import AudioToolbox

public class Sequencer {
  var processingGraph: AUGraph = nil
  var mixerUnit: AudioUnit = nil
  var mixerNode: AUNode = 0
  var ioNode: AUNode = 0
  var musicPlayer: MusicPlayer = nil
  var musicSequence: MusicSequence = nil
  var tune: Tune! = nil

  public init() {
    prepareGraph()
    checkError(NewMusicSequence(&musicSequence))
    checkError(MusicSequenceSetAUGraph(musicSequence, processingGraph))
    checkError(NewMusicPlayer(&musicPlayer))
    checkError(MusicPlayerSetSequence(musicPlayer, musicSequence))

    // MusicPlayer seems to play 1 musicstamp in 0.5 sec.
    checkError(MusicPlayerSetPlayRateScalar(musicPlayer, 0.5));
  }

  public func loadTune(tune: Tune) {
    self.tune = tune

    for (i, voice) in tune.tuneBody.voices.enumerate() {
      var samplerNode: AUNode = 0
      var desc = AudioComponentDescription(
        componentType: kAudioUnitType_MusicDevice,
        componentSubType: kAudioUnitSubType_Sampler,
        componentManufacturer: kAudioUnitManufacturer_Apple,
        componentFlags: 0,
        componentFlagsMask: 0)
      checkError(AUGraphAddNode(processingGraph, &desc, &samplerNode))
      checkError(AUGraphConnectNodeInput(processingGraph, samplerNode, 0, mixerNode, UInt32(i)))

      var unit: AudioUnit = nil
      checkError(AUGraphNodeInfo(processingGraph, samplerNode, nil, &unit))
      loadSF2Preset(0, unit: unit)
      loadVoice(voice, node: samplerNode)
    }

    var isInitialized: DarwinBoolean = false
    checkError(AUGraphIsInitialized(processingGraph, &isInitialized))
    if (!isInitialized) {
      checkError(AUGraphInitialize(processingGraph))
    }
    checkError(MusicPlayerPreroll(musicPlayer))
  }

  public func play() {
    var playing: DarwinBoolean = false
    checkError(MusicPlayerIsPlaying(musicPlayer, &playing))
    if (!playing) {
      checkError(MusicPlayerStart(musicPlayer))
    }
  }

  public func stop() {
    var playing: DarwinBoolean = false
    checkError(MusicPlayerIsPlaying(musicPlayer, &playing))
    if (playing) {
      checkError(MusicPlayerStop(musicPlayer))
    }
  }

  private func prepareGraph() {
    checkError(NewAUGraph(&processingGraph))
    checkError(AUGraphOpen(processingGraph))

    var mixerDesc = AudioComponentDescription(
      componentType: kAudioUnitType_Mixer,
      componentSubType: kAudioUnitSubType_MultiChannelMixer,
      componentManufacturer: kAudioUnitManufacturer_Apple,
      componentFlags: 0,
      componentFlagsMask: 0)
    checkError(AUGraphAddNode(processingGraph, &mixerDesc, &mixerNode))
    checkError(AUGraphNodeInfo(processingGraph, mixerNode, nil, &mixerUnit))
    checkError(AUGraphAddRenderNotify(processingGraph, renderCallback, nil))

    var ioDesc = AudioComponentDescription(
      componentType: kAudioUnitType_Output,
      componentSubType: kAudioUnitSubType_RemoteIO,
      componentManufacturer: kAudioUnitManufacturer_Apple,
      componentFlags: 0,
      componentFlagsMask: 0)
    checkError(AUGraphAddNode(processingGraph, &ioDesc, &ioNode))

    checkError(AUGraphConnectNodeInput(processingGraph, mixerNode, 0, ioNode, 0))
  }

  private func loadVoice(voice: Voice, node: AUNode) {
    var track: MusicTrack = nil
    checkError(MusicSequenceNewTrack(musicSequence, &track))
    let unitDuration = calculateUnitDuration(tune.tuneHeader.tempo)

    var timestamp: MusicTimeStamp = 0
    for elem in voice.elements {
      switch elem {
      case let n as Note:
        let note = pitchToMIDINote(tune.tuneHeader.key, pitch: n.pitch)
        let duration = calculateDuration(n, unitDuration: unitDuration)
        var message = defaultMIDINoteMessage(note, duration: duration)

        checkError(MusicTrackNewMIDINoteEvent(track, timestamp, &message))
        timestamp += MusicTimeStamp(duration)
      case let c as Chord:
        let duration = calculateDuration(c, unitDuration: unitDuration)
        for p in c.pitches {
          let note = pitchToMIDINote(tune.tuneHeader.key, pitch: p)
          var message = defaultMIDINoteMessage(note, duration: duration)

          checkError(MusicTrackNewMIDINoteEvent(track, timestamp, &message))
        }
        timestamp += MusicTimeStamp(duration)
      case let r as Rest:
        timestamp += MusicTimeStamp(calculateDuration(r, unitDuration: unitDuration))
      case let mr as MultiMeasureRest:
        timestamp += MusicTimeStamp(calculateDuration(tune.tuneHeader.meter, rest: mr, unitDuration: unitDuration))
      case let t as Tuplet:
        let f = Float32(t.notes) / Float32(t.time)
        for e in t.elements {
          let duration = calculateDuration(e, unitDuration: unitDuration) * f
          switch e {
          case let n as Note:
            let note = pitchToMIDINote(tune.tuneHeader.key, pitch: n.pitch)
            var message = defaultMIDINoteMessage(note, duration: duration)

            checkError(MusicTrackNewMIDINoteEvent(track, timestamp, &message))
          case let c as Chord:
            for p in c.pitches {
              let note = pitchToMIDINote(tune.tuneHeader.key, pitch: p)
              var message = defaultMIDINoteMessage(note, duration: duration)

              checkError(MusicTrackNewMIDINoteEvent(track, timestamp, &message))
            }
          default: ()
          }
          timestamp += MusicTimeStamp(duration)
        }
      default: ()
      }
    }

    checkError(MusicTrackSetDestNode(track, node))
  }
}

let renderCallback: AURenderCallback = { (inRefCon, ioActionFlags, inTimeStamp, inBusNumber, inNumberFrames, ioData) -> OSStatus in

  print(inTimeStamp.memory.mSampleTime)
  return noErr
}

func loadSF2Preset(preset: UInt8, unit: AudioUnit) {
  if let bankURL = NSBundle(forClass: Sequencer.self).URLForResource("vibraphone", withExtension: "sf2") {
    var instdata = AUSamplerInstrumentData(
      fileURL: Unmanaged.passUnretained(bankURL),
      instrumentType: UInt8(kInstrumentType_DLSPreset),
      bankMSB: UInt8(kAUSampler_DefaultMelodicBankMSB),
      bankLSB: UInt8(kAUSampler_DefaultBankLSB),
      presetID: preset)

    checkError(AudioUnitSetProperty(unit,
      kAUSamplerProperty_LoadInstrument,
      kAudioUnitScope_Global,
      0,
      &instdata,
      UInt32(sizeof(AUSamplerInstrumentData))))
  }
}

func checkError(status: OSStatus) -> Void {
  assert(status == noErr, "error code : \(status)")
}